clouddriver-ecs/src/main/java/com/netflix/spinnaker/clouddriver/ecs/view/EcsApplicationProvider.java

```diff
@@ -0,0 +1,147 @@
+import com.netflix.spinnaker.clouddriver.security.AccountCredentials;
+import com.netflix.spinnaker.clouddriver.security.AccountCredentialsProvider;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.stereotype.Component;
+
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.Map;
+import java.util.Set;
+
+@Component
+public class EcsApplicationProvider implements ApplicationProvider {
+
+  public static final String SERVICE_DELIMITER = "-";
+
+  AccountCredentialsProvider accountCredentialsProvider;
+  AmazonClientProvider amazonClientProvider;
+
+  @Autowired
+  public EcsApplicationProvider(AccountCredentialsProvider accountCredentialsProvider, AmazonClientProvider amazonClientProvider) {
+    this.accountCredentialsProvider = accountCredentialsProvider;
+    this.amazonClientProvider = amazonClientProvider;
+    getApplications(false);
+  }
+
+
+  @Override
+  public Application getApplication(String name) {
+
+    for (Application application: getApplications(false)) {
+      if (name.equals(application.getName())) {
+        return application;
+      }
+    }
+
+    return null;
+  }
+
+  @Override
+  public Set<Application> getApplications(boolean expand) {
+    Set<Application> applications = new HashSet<>();
+
+    for (AccountCredentials credentials: accountCredentialsProvider.getAll()) {
+      if (credentials instanceof AmazonCredentials) {
+        Set<Application> retrievedApplications = findApplicationsForAllRegions((AmazonCredentials) credentials);
+        applications.addAll(retrievedApplications);
+      }
+    }
+
+    return applications;
+  }
+
+  private Set<Application> findApplicationsForAllRegions(AmazonCredentials credentials) {
+    Set<Application> applications = new HashSet<>();
+
+    for (AmazonCredentials.AWSRegion awsRegion: credentials.getRegions()) {
+      AmazonECS amazonECS = amazonClientProvider.getAmazonEcs(
+        credentials.getName(), credentials.getCredentialsProvider(), awsRegion.getName());
+
+      applications.addAll(findApplicationsForRegion(amazonECS));
+    }
+
+    return applications;
+  }
+
+  private Set<Application> findApplicationsForRegion(AmazonECS amazonECS) {
+    HashMap<String, Application> applicationHashMap = populateApplicationMap(amazonECS);
+    Set<Application> applications = transposeApplicationMapToSet(applicationHashMap);
+
+    return applications;
+  }
+
+  private HashMap<String, Application> populateApplicationMap(AmazonECS amazonECS) {
+    HashMap<String, Application> applicationHashMap = new HashMap<>();
+    for (String clusterArn: amazonECS.listClusters().getClusterArns()) {
+      String clusterName = inferClusterNameFromArn(clusterArn);
+
+      ListServicesResult result = amazonECS.listServices(new ListServicesRequest().withCluster(clusterName));
+      for (String serviceArn: result.getServiceArns()) {
+        inferApplicationFromServices(amazonECS, applicationHashMap, clusterName, serviceArn);
```

Comment:
applicationHashMap is not modified in the method:

Does inferApplicationFromServices modify applicationHashMap? If so that seems like a code smell that you are modifying the parameter specially since the applications has a return type of HashMap<String, Application> that is being thrown away here.

Either:
1) Document that applicationHashMap parameter is expected to be modified, change the return Type to void
2) Return a modified hash and store it here in the applicationHashMap variable, and don't modify the parameter

Reply:
applicationHashMap was being modified by `inferApplicationFromServices`, but that was not clear.  Indeed, the method should be eplicit about it